### DISCUSSION TITLE
Support for Parameterized Target Markdown?

### DISCUSSION THREAD

**Discussion by 33Vito** [ropensci/targets#595] [Ideas]
## Prework

* [x] Read and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and [contributing guidelines](https://github.com/ropensci/targets/blob/main/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/ropensci/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [x] New features take time and effort to create, and they take even more effort to maintain. So if the purpose of the feature is to resolve a struggle you are encountering personally, please consider first posting a "trouble" or "other" issue so we can discuss your use case and search for existing solutions first.
* [x] Format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Proposal

Wonderful package, love it!

Apologies if I missed it, can I propose the support of parameterized Target Markdown? It would be a perfect match for scheduled Rmarkdown in RStudio Connect. 

I know `tar_render()` already supports parameterized R markdown, but the Target Markdown using the `targets` engine cannot pass on the markdown parameters `params` into the environment for the Target pipeline. 

Also keen if there is a workaround in the meanwhile. My current approach is for the Target Markdown to load an external yaml file to the `targets` chunk where `tar_globals = TRUE`.

Thanks! 
Tony

**Comment by wlandau** [comment]
Interesting idea. Your current workaround sounds good, as long as we are not talking about the special `_targets.yaml` file for `tar_config_set()`. If I understand correctly, you use the ordinary parameterized report mechanism, save the parameters to special file using an ordinary R code chunk, and then have a `{targets ..., tar_globals = TRUE}` code chunk to read that file. Sounds like it should work well.

I would prefer not to build something special into `targets`, however. To implement it safely, we would need to capture the code that generates each R Markdown parameter so we can insert valid R code into the text of each `{targets}` code chunk. But by the time the report is running, all the parameters are already evaluated and it is no longer possible to access the code that generated them. Some kinds of R objects cannot be reverse-engineered to produce the generating code. Others are just unpleasant to look at even if it works.

``` r
# Code cannot be reverse-engineered:
parameter_we_want <- new.env()
code <- deparse(parameter_we_want)
print(parameter_we_want)
#> <environment: 0x7fd17bbb7390>
print(code)
#> [1] "<environment>"
eval(parse(text = code))
#> Error in parse(text = code): <text>:1:1: unexpected '<'
#> 1: <
#>     ^

# Code does not look nice:
parameter_we_want <- Sys.Date()
code <- deparse(parameter_we_want)
print(parameter_we_want)
#> [1] "2021-08-11"
print(code)
#> [1] "structure(18850, class = \"Date\")"
eval(parse(text = code))
#> [1] "2021-08-11"
```

<sup>Created on 2021-08-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Reply by 33Vito** [reply]
Thanks, Will. I understand your concerns now. Because the parameters in the Rmarkdown can accept arbitrary R code, to deparse them become a challenge. 

My intention is to make it possible for end-user to alter the report parameters with the native Parameterized Rmarkdown interface (shiny) with out having to change the code. I have a better workaround now - I have a piece of R code at the top of the Rmarkdown which reads in the parameters and output them to a yaml file, then the `{targets ..., tar_globals = TRUE}` chunk can just consume the parameters from that yaml - its nice that way because we can also record the time when the report is run and what parameters are set in the yaml :)
